test/e2e/query_frontend_test.go

```diff
@@ -0,0 +1,193 @@
+			MaxSize:      "2KB",
+			MaxSizeItems: 1000,
+			Validity:     time.Hour,
+		},
+	}
+
+	queryFrontend, err := e2ethanos.NewQueryFrontend("1", "http://"+q.NetworkHTTPEndpoint(), inMemoryCacheConfig)
+	testutil.Ok(t, err)
+	testutil.Ok(t, s.StartAndWaitReady(queryFrontend))
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	t.Cleanup(cancel)
+
+	testutil.Ok(t, q.WaitSumMetricsWithOptions(e2e.Equals(1), []string{"thanos_store_nodes_grpc_connections"}, e2e.WaitMissingMetrics))
+
+	// Ensure we can get the result from Querier first so that it
+	// doesn't need to retry when we send queries to the frontend later.
+	queryAndAssertSeries(t, ctx, q.HTTPEndpoint(), queryUpWithoutInstance, promclient.QueryOptions{
+		Deduplicate: false,
+	}, []model.Metric{
+		{
+			"job":        "myself",
+			"prometheus": "test",
+			"replica":    "0",
+		},
+	})
+
+	now := time.Now()
+
+	t.Run("query frontend works for instant query", func(t *testing.T) {
+		queryAndAssertSeries(t, ctx, queryFrontend.HTTPEndpoint(), queryUpWithoutInstance, promclient.QueryOptions{
+			Deduplicate: false,
+		}, []model.Metric{
+			{
+				"job":        "myself",
+				"prometheus": "test",
+				"replica":    "0",
+			},
+		})
+
+		testutil.Ok(t, queryFrontend.WaitSumMetricsWithOptions(
+			e2e.Equals(1),
+			[]string{"thanos_query_frontend_queries_total"},
+			e2e.WithLabelMatchers(labels.MustNewMatcher(labels.MatchEqual, "op", "query"))),
+		)
+	})
+
+	t.Run("query frontend works for labels APIs", func(t *testing.T) {
+		// LabelNames and LabelValues API should still work via query frontend.
+		labelNames(t, ctx, queryFrontend.HTTPEndpoint(), timestamp.FromTime(now.Add(-time.Hour)), timestamp.FromTime(now.Add(time.Hour)), func(res []string) bool {
+			return len(res) > 0
+		})
+		labelValues(t, ctx, queryFrontend.HTTPEndpoint(), "instance", timestamp.FromTime(now.Add(-time.Hour)), timestamp.FromTime(now.Add(time.Hour)), func(res []string) bool {
+			return len(res) > 0
+		})
+	})
+
+	t.Run("query frontend works for range query and it can cache results", func(t *testing.T) {
+		rangeQuery(
+			t,
+			ctx,
+			queryFrontend.HTTPEndpoint(),
+			queryUpWithoutInstance,
+			timestamp.FromTime(now.Add(-time.Hour)),
+			timestamp.FromTime(now.Add(time.Hour)),
+			14,
+			promclient.QueryOptions{},
+			func(res model.Matrix) bool {
+				return len(res) > 0
+			},
+		)
+
+		testutil.Ok(t, queryFrontend.WaitSumMetricsWithOptions(
+			e2e.Equals(1),
+			[]string{"thanos_query_frontend_queries_total"},
+			e2e.WithLabelMatchers(labels.MustNewMatcher(labels.MatchEqual, "op", "query_range"))),
+		)
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(1), "cortex_cache_fetched_keys"))
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(0), "cortex_cache_hits"))
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(1), "querier_cache_added_new_total"))
+		testutil.Ok(t, queryFrontend.WaitSumMetrics(e2e.Equals(1), "querier_cache_added_total"))
```

Comment:
can we check if backend (querier) was touched? (: using some grpc_client conn metric on querier (: WDYT?

Reply:
Good idea! Let me update it.

Reply:
What about using `grpc_server_handled_total{method="xxx"}`?

Reply:
I think maybe checking `http_requests_total` on querier is also fine?

Reply:
yup :+1: